### PR TITLE
Show stack on errors

### PR DESCRIPTION
**Pull request by d3dave**

### CODE REVIEW

File: examples/cpp/pyperf/PyPerfCollapsedPrinter.cc [hunk 1]
```diff
@@ -78,41 +77,40 @@ void PyPerfCollapsedPrinter::processSamples(
...
+      } else {
+        std::fprintf(output_file, ";%s", kLostSymbol.c_str());
+        lostSymbols++;
+      }
+    }
+
     switch (sample.stackStatus) {
     case STACK_STATUS_TRUNCATED:
-      std::fprintf(output_file, ";%s", kIncompleteStack.c_str());
+      std::fprintf(output_file, ";%s", kTruncatedStack.c_str());
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfCollapsedPrinter.cc:95]: These 2 should be printed before the stacks themselves

**Comment by Jongy** [comment]
Since we now print stacks also in the error case, we need to move the initialization earlier. There's a bunch of stuff in `get_thread_state`, like:
```
  // Initialize stack info in case any subprogram below fails
  event->stack_status = STACK_STATUS_ERROR;
  event->stack_len = 0;
```

additionally there's the kernel stacks collection in `read_python_stack`:
```
event->kernel_stack_id = kernel_stacks.get_stackid(ctx, BPF_F_REUSE_STACKID);
```

we now need to set it to `-1` (or some other marker) so we don't show stale stacks in the error case.

File: examples/cpp/pyperf/PyPerfCollapsedPrinter.cc (hunk 1 above)
Comment (d3dave) [reply examples/cpp/pyperf/PyPerfCollapsedPrinter.cc]: Done

File: examples/cpp/pyperf/PyPerfBPFProgram.cc [hunk 2]
```diff
@@ -289,6 +289,11 @@ on_event(struct pt_regs* ctx) {
   event->tid = (pid_t)pid_tgid;
   bpf_get_current_comm(&event->comm, sizeof(event->comm));
 
+  // Initialize stack info
+  event->kernel_stack_id = kernel_stacks.get_stackid(ctx, BPF_F_REUSE_STACKID);
+  event->stack_status = STACK_STATUS_ERROR;
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfBPFProgram.cc:294]: Worth to initialize `event->error_code = ERROR_NONE` or `ERROR_UNSET`? To catch any case were we submit an event without properly initializing the error code (thus, using the previous one...)

Comment (d3dave) [reply examples/cpp/pyperf/PyPerfBPFProgram.cc]: There's no flow that allows that, but sure.

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfBPFProgram.cc]: Protecting from future bugs

File: examples/cpp/pyperf/PyPerfBPFProgram.cc [hunk 3]
```diff
@@ -291,8 +291,9 @@ on_event(struct pt_regs* ctx) {
 
   // Initialize stack info
   event->kernel_stack_id = kernel_stacks.get_stackid(ctx, BPF_F_REUSE_STACKID);
-  event->stack_status = STACK_STATUS_ERROR;
   event->stack_len = 0;
+  event->stack_status = STACK_STATUS_ERROR;
+  event->error_code = ERROR_NONE;
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfBPFProgram.cc:296]: So if we see `ERROR_NONE` from now on, it's a bug.

**Pull request merged**